### PR TITLE
BUG: Fixed merging on tz-aware

### PR DESCRIPTION
**Pull request by TomAugspurger**
Closes https://github.com/pandas-dev/pandas/issues/25014

### CODE REVIEW

File: pandas/core/internals/concat.py [hunk 1]
```diff
@@ -219,6 +219,11 @@ def get_reindexed_values(self, empty_dtype, upcasted_na):
 
         else:
             for ax, indexer in self.indexers.items():
+                # GH-25014: get_upcasted_na returns iNaT, but
+                # DatetimeArray.take expects NaT.
+                # TODO: update get_empty_dtype_and_na to use EAs earlier?
+                if is_datetime64tz_dtype(values) and fill_value == tslibs.iNaT:
```
Comment (jreback) [inline pandas/core/internals/concat.py]: isnull(fill_value)?

Comment (jreback) [reply pandas/core/internals/concat.py]: actually this should be in take_nd I think (instead)

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/internals/concat.py (hunk 1 above)
Comment (TomAugspurger) [reply pandas/core/internals/concat.py]: I was (hopefully) able to fix it at the root, in `get_empty_dtype_and_na`.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
lgtm. ping on green.

**Comment by TomAugspurger** [comment]
All green, but had to make an additional change in 
https://github.com/pandas-dev/pandas/pull/25033/commits/43706fb512ac75afd7999ad2198d5f08dfde7ce8

**Review by jorisvandenbossche — APPROVED** [review]

**Comment by jreback** [comment]
thanks @TomAugspurger

**Pull request merged**